CookieMediator/src/main/java/com/mitra/creatiocrm/headers/BPMCSRFCookie.java

```diff
@@ -0,0 +1,82 @@
+package com.mitra.creatiocrm.headers;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.apache.synapse.mediators.AbstractMediator;
+import org.apache.synapse.MessageContext;
+
+public class BPMCSRFCookie extends AbstractMediator {
+
+    private static final Log log = LogFactory.getLog(BPMCSRFCookie.class);
+    private static String COOKIE_NAME = "set-cookie";
+    private static String COOKIE_NAME_CAMELCASE = "Set-Cookie";
+    private static String COOKIE_BPMCSRF = "BPMCSRF=";
+    private static String AUTH_COMPLETE_PROPERTY = "authCompleteCookie";
+    private static String BPMSESSIONID = "BPMSESSIONID";
+    private static String USERNAME = "UserName";
+    private static String BPMLOADER = "BPMLOADER";
+    private static String ASPXAUTH = ".ASPXAUTH";
+    private static String BPMCSRF = "BPMCSRF";
+
+    public boolean mediate(MessageContext context) {
+
+        String allCookies = "";
+        Map<String, String> cookiesMap = new HashMap<String, String>();
+        Object headers = ((Axis2MessageContext) context)
+                .getAxis2MessageContext().getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+        Map<String, String> headerMap = (Map) headers;
+        if (headerMap.containsKey(COOKIE_NAME_CAMELCASE) ) {
+            String val = headerMap.get(COOKIE_NAME_CAMELCASE);
+            headerMap.remove(COOKIE_NAME_CAMELCASE);
+            headerMap.put(COOKIE_NAME, val);
+        }
+        if (headerMap.containsKey(COOKIE_NAME) ) {
+            String cookies = headerMap.get(COOKIE_NAME);
+            String authCookiesVal = cookies.split(";")[0];
+            if (cookies.contains(COOKIE_BPMCSRF)) {
+                context.setProperty(BPMCSRF, authCookiesVal.split(":")[1]);
+            }
+            cookiesMap.put( authCookiesVal.split("=")[0], cookies.split(";")[0]);
+            headerMap.remove(COOKIE_NAME);
+        }
+        Map<String, List<String>> excessHeadersMap = (Map) ((Axis2MessageContext) context).getAxis2MessageContext().getProperty("EXCESS_TRANSPORT_HEADERS");
+        if (excessHeadersMap.containsKey(COOKIE_NAME_CAMELCASE)) {
+            List<String> val = excessHeadersMap.get(COOKIE_NAME_CAMELCASE);
+            excessHeadersMap.remove(COOKIE_NAME_CAMELCASE);
```

Comment:
Why we can't simply get the cookie value from excess headers from the corresponding header name? Is that necessary to remove and update excess header map?

Reply:
Headers values are case sensitive and cookie header present as set-cookie or Set-Cookie, so above logic support for both way of presenting cookie value. When "Set-Cookie" header available,cookie update as 'set-cookie' and rest of the logic will be common for both case handling. Here, not necessary to remove, but it will add extra header if 'Set-Cookie' header present and not removed from the Map.